### PR TITLE
Asymmetric PCGrad: Protect In-Dist Gradients, Project OOD Only

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current 2-way PCGrad baseline (PR #2130) uses **symmetric** gradient surgery: when in-dist gradients (g_A) and OOD gradients (g_B) conflict, BOTH are projected to remove the conflicting component. This means even the in-dist gradient gets modified when it conflicts with OOD — sacrificing in-dist accuracy for the sake of balance.

**Asymmetric PCGrad** inverts this: only project g_B (OOD) onto the normal plane of g_A (in-dist). The in-dist gradient is never touched. This matches our actual problem structure: in-dist (p_in, p_oodc, p_re) is the anchor task; OOD tandem transfer (p_tan) is the adapting task that should learn without destabilizing the anchor.

**Expected impact:** -1% to -3% across OOD metrics (p_oodc, p_tan, p_re) while holding or improving p_in. The symmetry change is a 3-line code modification inside the existing PCGrad block.

**Why not tried before:** Previous PCGrad experiments (#2147) changed the task grouping (3-way vs 2-way). This changes the *projection direction* within the existing 2-way setup — a fundamentally different axis.

### Code change: `cfd_tandemfoil/train.py`

**Step 1: Add config flag** (~line 950 in the config dataclass):
```python
pcgrad_asymmetric: bool = False
```

**Step 2: Modify the PCGrad projection block** (~lines 1996-2010, inside the per-parameter loop where `dot_ab < 0`).

Find the current symmetric block:
```python
elif dot_ab < 0:
    p.grad = ((ga - (dot_ab / gb_ns) * gb) + (gb - (dot_ab / ga_ns) * ga)) * 0.5
else:
    p.grad = (ga + gb) * 0.5
```

Change it to:
```python
elif dot_ab < 0:
    if cfg.pcgrad_asymmetric:
        # ASYMMETRIC: only project g_B (OOD) onto normal plane of g_A (in-dist)
        # g_A is preserved; g_B has its g_A-conflicting component removed
        gb_proj = gb - (dot_ab / ga_ns) * ga
        p.grad = (ga + gb_proj) * 0.5
    else:
        # SYMMETRIC (original): both gradients projected
        p.grad = ((ga - (dot_ab / gb_ns) * gb) + (gb - (dot_ab / ga_ns) * ga)) * 0.5
else:
    p.grad = (ga + gb) * 0.5
```

**Step 3: Wire the flag** — Add `--pcgrad_asymmetric` as a `store_true` argparse flag and pass it through to cfg.

**Step 4: Run 2 seeds:**

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent edward \
  --wandb_name "edward/asymmetric-pcgrad-s42" --wandb_group phase6/asymmetric-pcgrad \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias \
  --pcgrad_asymmetric

# Seed 73 (same flags, --seed 73, --wandb_name "edward/asymmetric-pcgrad-s73")
```

### Important implementation note

The `ga_ns = (ga_flat @ ga_flat).item()` in the current code is the GLOBAL gradient norm squared (computed once outside the per-parameter loop). The asymmetric formula uses `ga_ns` as the denominator — use the same pre-computed value.

For single-foil samples where only one task gradient exists (g_B is zero), the `dot_ab < 0` branch will not fire — behavior is identical to the symmetric baseline for single-foil batches.

### What to report

| Config | Seed | p_in | p_oodc | p_tan | p_re | W&B |
|--------|------|------|--------|-------|------|-----|
| asymmetric-pcgrad | 42 | | | | | |
| asymmetric-pcgrad | 73 | | | | | |
| **avg** | — | | | | | |
| **Baseline (symmetric, lr=2e-4)** | — | 13.05 | 7.70 | 28.60 | 6.55 | d7l91p0x, j9btfx09 |

Also report: did p_in hold steady or improve vs baseline? (key signal that asymmetric projection is working correctly)

## Baseline

Current baseline (PR #2130, GSB + PCGrad symmetric, lr=2e-4, 2-seed avg):

| Metric | 2-seed avg | Target |
|--------|-----------|--------|
| p_in   | 13.05     | < 13.05 |
| p_oodc | 7.70      | < 7.70  |
| **p_tan** | **28.60** | **< 28.60** |
| p_re   | 6.55      | < 6.55  |

W&B: d7l91p0x (seed 42), j9btfx09 (seed 73)

```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-gsb-pcgrad" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results: Asymmetric PCGrad

### Setup
- 2 runs: seeds 42 and 73
- All flags match baseline + `--pcgrad_asymmetric`
- Implementation: when in-dist (g_A) and OOD (g_B) gradients conflict (dot < 0), only project g_B onto the normal plane of g_A. g_A is preserved unchanged.
- Both runs reached epoch 146-147 before 180-min timeout

### Results — Surface MAE (Pressure)

| Run | p_in | p_oodc | p_tan | p_re | val/loss | W&B Run ID |
|-----|------|--------|-------|------|----------|------------|
| Asymmetric s42 | 13.3 | 7.7 | **28.9** | 6.5 | 0.387 | `amb3ci1h` |
| Asymmetric s73 | 13.5 | 8.0 | **28.8** | 6.5 | 0.386 | `hc1vp76q` |
| **Asymmetric avg** | **13.4** | **7.85** | **28.85** | **6.50** | — | — |
| **Baseline (symmetric)** | **13.05** | **7.70** | **28.60** | **6.55** | — | — |

### Comparison vs Baseline

| Metric | Asymmetric | Baseline | Delta |
|--------|-----------|----------|-------|
| p_in | 13.4 | 13.05 | +2.7% |
| p_oodc | 7.85 | 7.70 | +1.9% |
| p_tan | 28.85 | 28.60 | +0.9% |
| p_re | 6.50 | 6.55 | **-0.8%** |

### Analysis

**Near-baseline performance, slightly worse overall.** The asymmetric variant doesn't beat baseline on any metric except p_re (marginal -0.8%).

**Key observations:**
- **p_in regressed by 2.7%** — this is the opposite of what was expected. Protecting g_A (in-dist) should have held or improved p_in. The regression suggests that the symmetric projection was actually beneficial for in-dist: projecting away the conflicting OOD component from g_A was acting as implicit regularization.
- **p_tan at 28.85 is competitive** but doesn't beat baseline's 28.60. The seed variance is low (28.8-28.9), suggesting the result is reliable.
- **p_re slightly improved** (6.50 vs 6.55), the only metric that benefited.
- **val/loss is comparable** (0.386-0.387 vs baseline ~0.385).

**Why symmetric may be better than asymmetric here:**
The symmetric PCGrad removes conflicting components from BOTH gradients, which effectively regularizes both tasks. For in-dist, this means the optimizer doesn't chase solutions that hurt OOD — which paradoxically helps p_in by preventing overfitting to in-dist-specific patterns that don't generalize.

### Conclusion

**Symmetric PCGrad (current baseline) is slightly better.** The asymmetric variant doesn't provide the expected benefit. The symmetric projection's implicit regularization of the in-dist gradient appears beneficial.

### Suggested Follow-ups
1. **Close this direction** — symmetric PCGrad is already well-tuned for this problem.
2. If exploring further, try the **reverse asymmetric** (protect OOD, project in-dist) to confirm the symmetry is important.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2158 because ADVISOR: Closing — asymmetric PCGrad doesn't beat symmetric baseline.

**Results:** p_tan=28.85 (+0.9%), p_in=13.4 (+2.7%), p_oodc=7.85 (+1.9%). Only p_re marginally better (-0.8%).

**Important finding:** Protecting g_A (in-dist) from OOD interference actually HURT p_in (+2.7%). This reveals that symmetric PCGrad's implicit regularization — projecting away conflicting components from BOTH gradients — is beneficial. It prevents overfitting to in-dist-specific patterns that don't generalize. The symmetry is a feature, not a bug.

Symmetric 2-way PCGrad confirmed as the optimal gradient surgery configuration. Combined with prior results: 3-way (#2147) = worse, asymmetric (#2158) = slightly worse, symmetric 2-way = best.

Good analysis, edward — the insight about symmetric projection as implicit regularization is valuable. Moving you to a novel augmentation experiment..
